File: xtask/src/demo_builds.rs
use crate::utils::{npm_program, path_from_repo, repo_root, run_step};

const DEMOS: &[(&str, &str)] = &[
    ("ts-basic", "demo/ts/basic"),
    ("ts-variants", "demo/ts/variants"),
    ("ts-theme", "demo/ts/theme"),
    ("ts-workspace", "demo/ts/workspace"),
    ("react-basic", "demo/react/basic"),
    ("react-variants", "demo/react/variants"),
    ("react-theme", "demo/react/theme"),
    ("react-workspace", "demo/react/workspace"),
    ("vue-basic", "demo/vue/basic"),
    ("vue-variants", "demo/vue/variants"),
    ("vue-theme", "demo/vue/theme"),
    ("vue-workspace", "demo/vue/workspace"),
];

pub fn run() -> Result<(), String> {
    let root = repo_root();
    let npm = npm_program();

    for (name, relative_path) in DEMOS {
        let demo_dir = path_from_repo(&root, relative_path);
        run_step(
            &format!("{name}: generate motif.css"),
            "cargo",
            &["run", "-p", "motif-core", "--", "."],
            &demo_dir,
        )?;
        run_step(
            &format!("{name}: install dependencies"),
            npm,
            &["install", "--no-package-lock"],
            &demo_dir,
        )?;
        run_step(&format!("{name}: build"), npm, &["run", "build"], &demo_dir)?;
    }

    Ok(())
}

File: scripts/node/demo-builds.mjs
import { fileURLToPath } from 'node:url';
import { npmProgram, parseOption, pathFromRepo, repoRoot, runStep } from './common.mjs';

const scriptPath = fileURLToPath(import.meta.url);

const tsDemos = [
  ['ts-basic', 'demo/ts/basic'],
  ['ts-variants', 'demo/ts/variants'],
  ['ts-theme', 'demo/ts/theme'],
  ['ts-workspace', 'demo/ts/workspace'],
];

const webDemos = [
  ['react-basic', 'demo/react/basic'],
  ['react-variants', 'demo/react/variants'],
  ['react-theme', 'demo/react/theme'],
  ['react-workspace', 'demo/react/workspace'],
  ['vue-basic', 'demo/vue/basic'],
  ['vue-variants', 'demo/vue/variants'],
  ['vue-theme', 'demo/vue/theme'],
  ['vue-workspace', 'demo/vue/workspace'],
];

async function buildTsDemo(name, relativePath) {
  const demoDir = pathFromRepo(relativePath);
  await runStep(
    `${name}: generate motif.css`,
    'cargo',
    ['run', '-p', 'motif-core', '--', '.'],
    demoDir
  );
  await runStep(
    `${name}: install dependencies`,
    npmProgram,
    ['install', '--no-package-lock'],
    demoDir
  );
  await runStep(`${name}: build`, npmProgram, ['run', 'build'], demoDir);
}

async function buildWebDemo(name, relativePath) {
  const demoDir = pathFromRepo(relativePath);
  await runStep(
    `${name}: install dependencies`,
    npmProgram,
    ['install', '--no-package-lock'],
    demoDir
  );
  await runStep(`${name}: build`, npmProgram, ['run', 'build'], demoDir);
}

async function main() {
  const profile = parseOption('--profile', 'full');

  if (profile !== 'full' && profile !== 'fast') {
    throw new Error(`unknown demo-builds profile \`${profile}\``);
  }

  if (profile === 'full') {
    for (const [name, relativePath] of tsDemos) {
      await buildTsDemo(name, relativePath);
    }
  }

  for (const [name, relativePath] of webDemos) {
    await buildWebDemo(name, relativePath);
  }
}

main().catch((error) => {
  console.error(`demo-builds failed: ${error.message}`);
  console.error(`hint: rerun with \`node ${pathFromRepo('scripts/node/demo-builds.mjs')} --profile=full\``);
  process.exit(1);
});
